File: hedera-mirror-rest/model/contractLog.js
// SPDX-License-Identifier: Apache-2.0

import _ from 'lodash';

class ContractLog {
  /**
   * Parses contract_log table columns into object
   */
  constructor(contractLog, recordFile) {
    Object.assign(
      this,
      _.mapKeys(contractLog, (v, k) => _.camelCase(k))
    );
    this.blockHash = recordFile?.hash;
    this.blockNumber = recordFile?.index ?? null;
  }

  static tableAlias = 'cl';
  static tableName = 'contract_log';

  static BLOOM = 'bloom';
  static CONSENSUS_TIMESTAMP = 'consensus_timestamp';
  static CONTRACT_ID = 'contract_id';
  static DATA = 'data';
  static INDEX = 'index';
  static PAYER_ACCOUNT_ID = 'payer_account_id';
  static ROOT_CONTRACT_ID = 'root_contract_id';
  static TOPIC0 = 'topic0';
  static TOPIC1 = 'topic1';
  static TOPIC2 = 'topic2';
  static TOPIC3 = 'topic3';
  static TRANSACTION_HASH = 'transaction_hash';
  static TRANSACTION_INDEX = 'transaction_index';

  /**
   * Gets full column name with table alias prepended.
   *
   * @param {string} columnName
   * @private
   */
  static getFullName(columnName) {
    return `${this.tableAlias}.${columnName}`;
  }
}

export default ContractLog;

File: common/src/main/java/org/hiero/mirror/common/domain/contract/ContractLog.java
// SPDX-License-Identifier: Apache-2.0

package org.hiero.mirror.common.domain.contract;

import com.fasterxml.jackson.annotation.JsonIgnore;
import jakarta.persistence.Convert;
import jakarta.persistence.Entity;
import jakarta.persistence.IdClass;
import jakarta.persistence.Transient;
import java.io.Serializable;
import lombok.AccessLevel;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Data;
import lombok.EqualsAndHashCode;
import lombok.NoArgsConstructor;
import lombok.ToString;
import org.hiero.mirror.common.converter.EntityIdConverter;
import org.hiero.mirror.common.domain.entity.EntityId;
import org.hiero.mirror.common.util.LogsBloomFilter;
import org.springframework.data.domain.Persistable;

@AllArgsConstructor(access = AccessLevel.PRIVATE) // For Builder
@Builder
@Data
@EqualsAndHashCode(exclude = "contractResult")
@Entity
@IdClass(ContractLog.Id.class)
@NoArgsConstructor
public class ContractLog implements Persistable<ContractLog.Id> {

    @ToString.Exclude
    private byte[] bloom;

    @jakarta.persistence.Id
    private long consensusTimestamp;

    @Convert(converter = EntityIdConverter.class)
    private EntityId contractId;

    @ToString.Exclude
    private byte[] data;

    @jakarta.persistence.Id
    private int index;

    @Convert(converter = EntityIdConverter.class)
    private EntityId rootContractId;

    @Convert(converter = EntityIdConverter.class)
    private EntityId payerAccountId;

    private byte[] topic0;

    private byte[] topic1;

    private byte[] topic2;

    private byte[] topic3;

    private byte[] transactionHash;

    private Integer transactionIndex;

    private boolean synthetic;

    /**
     * Transient reference to the ContractResult this log belongs to.
     * Used for updating the bloom filter in the correct ContractResult during synthetic log processing.
     */
    @JsonIgnore
    @Transient
    @ToString.Exclude
    private ContractResult contractResult;

    @Override
    @JsonIgnore
    public Id getId() {
        Id id = new Id();
        id.setConsensusTimestamp(consensusTimestamp);
        id.setIndex(index);
        return id;
    }

    @JsonIgnore
    @Override
    public boolean isNew() {
        return true;
    }

    public void setBloom(final byte[] bloom) {
        if (bloom == null) {
            return;
        }

        this.bloom = bloom;
        if (synthetic && contractResult != null) {
            final var existingResultBloom = contractResult.getBloom();
            final var aggregatedBloom = bloom.length == LogsBloomFilter.BYTE_SIZE
                    ? LogsBloomFilter.or(existingResultBloom, bloom)
                    : existingResultBloom;

            contractResult.setBloom(aggregatedBloom);
        }
    }

    @Data
    @AllArgsConstructor
    @NoArgsConstructor
    public static class Id implements Serializable {
        private static final long serialVersionUID = -6192177810161178246L;
        private long consensusTimestamp;
        private int index;
    }
}
